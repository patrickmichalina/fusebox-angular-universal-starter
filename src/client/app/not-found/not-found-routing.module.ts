import { NotFoundComponent } from './not-found.component'
import { NgModule } from '@angular/core'
import { RouterModule } from '@angular/router'
import { MetaGuard } from '@ngx-meta/core'

@NgModule({
  imports: [
    RouterModule.forChild([
      {
        path: '**',
        component: NotFoundComponent,
        canActivate: [MetaGuard],
        data: {
          meta: {
            title: 'i18n.not-found.title',
            description: 'i18n.not-found.description'
          }
        }
      }
    ])
  ],
  exports: [RouterModule]
})
export class NotFoundRoutingModule { }
