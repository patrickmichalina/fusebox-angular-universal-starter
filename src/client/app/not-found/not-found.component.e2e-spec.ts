import { baseUrl, browser } from '../../../../tools/test/jest.e2e-setup'

describe('NotFound Page', () => {
  it('should have title', async () => {
    const page = browser.goto(`${baseUrl}/not-found`)

    const text = await page.evaluate(() => document.title)

    expect(text).toContain('Not Found - Fusebox Angular Universal Starter')
  })
})
