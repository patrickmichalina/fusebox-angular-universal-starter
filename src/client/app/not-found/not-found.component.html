<md-card>
  PAGE NOT FOUND
</md-card>