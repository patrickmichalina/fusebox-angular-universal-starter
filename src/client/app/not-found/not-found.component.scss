:host {
  .footer {
    text-align: right;
    padding: .75em;
  }
   ::ng-deep mat-tab-group {
    flex: 1;
    height: 100%;
    .mat-tab-body-wrapper {
      height: 100%;
    }
    mat-tab-body {
      .mat-tab-body-content {
        overflow: overlay;
      }
      // overflow-y: overlay;
      // .mat-tab-body-content {
      //   overflow-y: overlay;
      // }
    }
  }
  .settings {
    display: flex;
    flex-direction: column;
    padding: 1em;
    height: 100%;
    overflow-y: auto;
  }
   ::ng-deep dynamic-html {
    display: -webkit-inline-box;
    overflow-y: auto;
    height: 100%;
  }
   ::ng-deep mat-form-field {
    padding-bottom: 1em;
  }
  .example-headers-align .mat-expansion-panel-header-title,
  .example-headers-align .mat-expansion-panel-header-description {
    flex-basis: 0;
  }
  .example-headers-align .mat-expansion-panel-header-description {
    justify-content: space-between;
    align-items: center;
  }
  .field-spacer {
    display: flex;
    flex: 1;
    mat-form-field {
      width: 100%;
      margin-right: 1em;
    }
    @media only screen and (max-width: 959px) {
      flex-direction: column;
    }
  }
   ::ng-deep mat-chip {
    mat-icon {
      height: auto;
      width: auto;
      margin-left: 8px;
      padding-bottom: 4px;
      &:hover {
        cursor: pointer;
      }
    }
  }
}