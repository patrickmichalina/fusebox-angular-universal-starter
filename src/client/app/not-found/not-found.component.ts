import { ChangeDetectionStrategy, Component, HostBinding, ViewChild } from '@angular/core'
import { QuillEditorComponent } from './../shared/quill-editor/quill-editor.component'
import { FirebaseDatabaseService } from './../shared/services/firebase-database.service'
import { AuthService } from './../shared/services/auth.service'
import { FormControl, FormGroup, Validators } from '@angular/forms'
import { ServerResponseService } from './../shared/services/server-response.service'
import { ActivatedRoute, Router } from '@angular/router'
import { Observable } from 'rxjs/Observable'
import { SEONode, SEOService } from '../shared/services/seo.service'
import { MatChipInputEvent, MatDialog, MatSnackBar } from '@angular/material'
import { ModalConfirmationComponent } from '../shared/modal-confirmation/modal-confirmation.component'
// tslint:disable-next-line:no-require-imports
import ms = require('ms')
import { ENTER } from '@angular/cdk/keycodes'

const COMMA = 188

export interface Page {
  content: string
  title: string
  isDraft: boolean
  userCommentsEnabled?: boolean
  cache?: { [key: string]: boolean | string | number }
  imgWidth?: number,
  imgHeight?: number,
  imgAlt?: string,
  imgUrl?: string,
  imgMime?: string
  articleTag?: string[]
}

@Component({
  selector: 'pm-not-found',
  templateUrl: './not-found.component.html',
  styleUrls: ['./not-found.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class NotFoundComponent {
  @HostBinding('class.vert-flex-fill') flexFill = true
  @ViewChild(QuillEditorComponent) editor: QuillEditorComponent

  addOnBlur = true
  separatorKeysCodes = [ENTER, COMMA]
  tags: string[] = []

  add(event: MatChipInputEvent): void {
    if (event.input) event.input.value = '' // clear input value
    this.tags = [...this.tags, event.value.trim()]
      .filter(a => a !== '')
      .filter((elem, pos, arr) => arr.indexOf(elem) === pos)
    this.updateTags()
  }

  remove(tag: any): void {
    this.tags = [...this.tags].filter(a => a !== tag)
    this.updateTags()
  }

  updateTags() {
    this.settingsForm.controls['articleTag'].setValue(this.tags)
  }

  updateTabParam(tab: number) {
    this.router.navigate([this.router.url.split('?')[0]], { queryParams: { tab }, queryParamsHandling: 'merge' })
  }

  private params$ = this.ar.queryParams.shareReplay()

  private isEditMode$ = this.params$
    .map(a => a.edit ? true : false)

  private currentTab$ = this.params$
    .map(a => a.tab ? +a.tab : 0)

  private url$ = Observable.of(this.router.url.split('?')[0])
    .filter(a => !a.includes('.'))
    .shareReplay()

  public settingsForm = new FormGroup({
    title: new FormControl('', [
      Validators.required
    ]),
    description: new FormControl('', [
      Validators.required,
      Validators.max(158)
    ]),
    imgUrl: new FormControl('', [
      // Validators.required
    ]),
    imgAlt: new FormControl('', [
      // Validators.min(1)
    ]),
    imgMime: new FormControl('', [
      // Validators.min(1)
    ]),
    imgHeight: new FormControl('', [
      Validators.min(1)
    ]),
    imgWidth: new FormControl('', [
      Validators.min(1)
    ]),
    articlePublishedTime: new FormControl(new Date(), [
      // Validators.min(1)
    ]),
    articleModifiedTime: new FormControl('', [
      // Validators.min(1)
    ]),
    articleExpirationTime: new FormControl('', [
      // Validators.min(1)
    ]),
    articleAuthor: new FormControl('', [
      // Validators.min(1)
    ]),
    articleSection: new FormControl('', [
      // Validators.min(1)
    ]),
    articleTag: new FormControl('', []),
    userCommentsEnabled: new FormControl('', []),
    isDraft: new FormControl('', [])
  })

  public page$ = this.url$
    .flatMap(url => this.db
      .get<Page & SEONode>(`/pages/${url}`)
      .flatMap(page => this.isEditMode$, (page, editMode) => ({ page, editMode }))
      .map(res => {
        if (res && res.page && (res.editMode || !res.page.isDraft)) {
          if (!res.editMode) {
            const pageCacheSettings = res.page.cache || {}
            const cacheControl = Object.keys(pageCacheSettings)
              .filter(key => pageCacheSettings[key])
              .reduce((acc, curr) => {
                const ret = typeof pageCacheSettings[curr] === 'boolean'
                  ? curr
                  : typeof pageCacheSettings[curr] === 'string'
                    ? `${curr}=${ms(pageCacheSettings[curr] as string) / 1000}`
                    : `${curr}=${pageCacheSettings[curr]}`

                return acc.concat(', ').concat(ret)
              }, '')
              .replace(/(^,)|(,$)/g, '')
              .trim()

            cacheControl
              ? this.rs.setHeader('Cache-Control', cacheControl)
              : this.rs.setCacheNone()
          } else {
            this.rs.setCacheNone()
          }

          return {
            ...res.page,
            content: res.page.content
          }
        }
        this.rs.setNotFound()
        this.rs.setCacheNone()
        return {
          ...res.page,
          content: 'not found'
        } as Page & SEONode
      })
      .do(page => {
        this.seo.updateNode({
          title: page.title,
          description: page.description,
          img: {
            width: page.imgWidth,
            height: page.imgHeight,
            type: page.imgMime,
            alt: page.imgAlt,
            url: page.url
          },
          tags: page.articleTag
        })

        // tslint:disable:no-null-keyword
        const formValues = Object.keys(this.settingsForm.controls).reduce((acc: any, controlKey) => {
          acc[controlKey] = (page as any)[controlKey] || null
          return acc
        }, {})

        this.settingsForm.setValue(formValues)
        this.tags = page.articleTag || []
      })
      .catch(err => {
        if (err.code === 'PERMISSION_DENIED') {
          this.rs.setStatus(401)
          return Observable.of({
            content: 'unauthorized'
          })
        } else {
          this.rs.setError()
          return Observable.of({
            content: err || 'server error'
          })
        }
      }))

  view$ = Observable.combineLatest(this.auth.user$, this.page$, this.currentTab$,
    this.ar.queryParams.pluck('edit').map(a => a === 'true'),
    (user, page, currentTab, isEditing) => {
      return {
        currentTab,
        canEdit: true, // for demo only, user && user.roles && user.roles.admin,
        isEditing,
        page
      }
    })
    .catch(err => {
      this.rs.setError()
      return Observable.of({
        content: 'server error'
      })
    })

  publish() {
    const settings = Object.keys(this.settingsForm.value)
      .filter(key => typeof this.settingsForm.value[key] !== 'undefined')
      .reduce((acc, curr) => {
        const obj = { ...acc } as any
        obj[curr] = this.settingsForm.value[curr]
        return obj
      }, {})

    this.url$.flatMap(url => this.db.getObjectRef(`/pages/${url}`)
      .update({
        ...settings,
        content: this.editor.textValue.getValue()
      }), (url, update) => ({ url, update }))
      .take(1)
      .subscribe(a => {
        this.router.navigate([a.url])
        this.showSnack('Published! Page is now live.')
      })
  }

  showSnack(message: string) {
    this.snackBar.open(message, 'dismiss', {
      duration: 2000,
      horizontalPosition: 'left',
      verticalPosition: 'bottom'
    })
  }

  confirmDelete() {
    return this.dialog.open(ModalConfirmationComponent, {
      width: '460px',
      position: {
        top: '30px'
      },
      data: {
        message: 'Deleting this page will immediately remove it from the database and anyone reading it',
        title: 'Are you sure?'
      }
    })
  }

  delete() {
    this.confirmDelete()
      .afterClosed()
      .filter(Boolean)
      .flatMap(() => this.url$)
      .flatMap(url => this.db.getObjectRef(`/pages/${url}`).remove(), (url, update) => ({ url, update }))
      .take(1)
      .subscribe(a => {
        this.router.navigate(['/pages'])
        this.showSnack('Page removed!')
      })
  }

  viewCurrent() {
    this.url$.do(url => {
      this.router.navigate([url])
    }).take(1).subscribe()
  }

  edit() {
    this.url$.do(url => {
      this.router.navigate([url], { queryParams: { edit: true } })
    }).take(1).subscribe()
  }

  constructor(private rs: ServerResponseService, private db: FirebaseDatabaseService, private seo: SEOService,
    public auth: AuthService, private ar: ActivatedRoute, private router: Router, private dialog: MatDialog,
    private snackBar: MatSnackBar) {
  }
}
