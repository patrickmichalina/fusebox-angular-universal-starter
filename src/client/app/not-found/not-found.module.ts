import { ServerResponseService } from './../shared/services/server-response.service'
import { NotFoundRoutingModule } from './not-found-routing.module'
import { NotFoundComponent } from './not-found.component'
import { NgModule } from '@angular/core'
import { SharedModule } from '../shared/shared.module'

@NgModule({
  imports: [NotFoundRoutingModule, SharedModule],
  declarations: [NotFoundComponent],
  exports: [NotFoundComponent],
  providers: [ServerResponseService]
})
export class NotFoundModule { }
