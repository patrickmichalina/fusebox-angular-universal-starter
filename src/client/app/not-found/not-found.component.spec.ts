import { ServerResponseService } from './../shared/services/server-response.service';
import { NotFoundComponent } from './not-found.component';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { Component } from '@angular/core';
import { NotFoundModule } from './not-found.module';

describe(NotFoundComponent.name, () => {
  let fixture: ComponentFixture<NotFoundComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [NotFoundModule],
      declarations: [TestComponent],
      providers: [ServerResponseService]
    }).compileComponents();
  }));

  beforeEach(async(() => {
    fixture = TestBed.createComponent(NotFoundComponent);
  }));

  afterEach(async(() => {
    TestBed.resetTestingModule();
  }));

  it('should match snapshot', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should compile', async(() => {
    fixture.detectChanges();
    expect(fixture.componentInstance).toBeDefined();
  }));

  it('should show text', async(() => {
    fixture.detectChanges();
    expect(fixture.debugElement.nativeElement.innerHTML).toBe('PAGE NOT FOUND');
  }));
});

@Component({
  selector: 'test-component',
  template: '<pm-not-found></pm-not-found>'
})
class TestComponent {}
