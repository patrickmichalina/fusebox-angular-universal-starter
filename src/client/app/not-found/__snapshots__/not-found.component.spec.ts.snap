// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFoundComponent should match snapshot 1`] = `
<pm-not-found>
  <md-card
    class="mat-card"
  >
    
    PAGE NOT FOUND
  
  </md-card>
</pm-not-found>
`;

exports[`NotFoundComponent should show text 1`] = `
<pm-not-found>
  <md-card
    class="mat-card"
  >
    
    PAGE NOT FOUND
  
  </md-card>
</pm-not-found>
`;
